components/common/src/main/java/com/hotels/styx/common/content/FlowControllingHttpContentProducer.java

```diff
@@ -78,39 +90,46 @@
         TERMINATED
     }
 
-    FlowControllingHttpContentProducer(
+    public FlowControllingHttpContentProducer(
             Runnable askForMore,
             Runnable onCompleteAction,
             Consumer<Throwable> onTerminateAction,
-            Runnable delayedTearDownAction,
             String loggingPrefix,
-            Origin origin) {
+            Origin origin,
+            long inactivityTimeoutMs,
+            EventLoop eventLoop,
+            String uri) {
         this.askForMore = requireNonNull(askForMore);
         this.onCompleteAction = requireNonNull(onCompleteAction);
         this.onTerminateAction = requireNonNull(onTerminateAction);
-        this.delayedTearDownAction = requireNonNull(delayedTearDownAction);
-        this.origin = requireNonNull(origin);
-
+        this.loggingPrefix = loggingPrefix;
+        this.origin = origin;
+        this.uri = uri;
+        if (uri.startsWith("_")) {
+            fsmHistories.put(uri, "Initial state:BUFFERING");
+        }
         this.stateMachine = new StateMachine.Builder<ProducerState>()
                 .initialState(BUFFERING)
 
                 .transition(BUFFERING, RxBackpressureRequestEvent.class, this::rxBackpressureRequestInBuffering)
                 .transition(BUFFERING, ContentChunkEvent.class, this::contentChunkInBuffering)
+                .transition(BUFFERING, TearDownEvent.class, this::releaseAndTerminate)
                 .transition(BUFFERING, ChannelInactiveEvent.class, this::releaseAndTerminate)
```

Comment:
The timer is not cancelled in `releaseAndTerminate` transition. Is this correct? My gut feeling is that it should be cancelled (or stopped) in every transition to a terminal state.